astropy/cosmology/_src/utils.py

```diff
@@ -68,32 +68,51 @@ def wrapper(self, *args, **kwargs):
     return wrapper
 
 
+# ===================================================================
+
+ScalarTypes: TypeAlias = Number | np.generic
+SCALAR_TYPES: Final = (float, int, np.generic, Number)  # arranged for speed
+
+
 class HasShape(Protocol):
     shape: tuple[int, ...]
 
 
 def aszarr(
-    z: Number | np.generic | HasShape | ArrayLike,
-) -> Number | np.generic | NDArray[Any]:
-    """Redshift as a `~numbers.Number` or |ndarray| / |Quantity| / |Column|.
+    z: Quantity | NDArray[Any] | ArrayLike | ScalarTypes | HasShape, /
+) -> NDArray[Any]:
+    """Redshift as an Array duck type.
 
     Allows for any ndarray ducktype by checking for attribute "shape".
     """
-    if isinstance(z, (Number, np.generic)):  # scalars
+    # Scalars
+    if isinstance(z, SCALAR_TYPES):
+        return np.asarray(z)
+
+    # Arrays
+    # Need to special-case Quantity since it is an NDArray subclass
+    elif isinstance(z, Quantity):
+        return (z << cu.redshift).value  # for speed only use enabled equivs
```

Comment:
Note that `q.value` will return a scalar if `shape==()`. You'd want `.view(np.ndarray)` here or add an explicit `[...]`.

Aside, I think everything would be caught with,
```
return Quantity(z, cu.redshift, copy=False).view(np.ndarray)
```
But it might be slow for `ndarray` input, so you could special case just that.
if isinstance(z, np.ndarary) and not isinstance(

Reply:
Won't that also be slow for Quantity inputs? `(z << cu.redshift)` is the fast path, right?

Reply:
Done!